pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/intervalST/IntervalST.java

```diff
@@ -0,0 +1,187 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.segmentpruner.intervalST;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+
+// A read-only interval search tree based on AVL Tree
+public class IntervalST<Value> {
+
+  private Node<Value> root;
+
+  public IntervalST(List<Long> mins, List<Long> maxs, List<Value> values) {
+    // read only once constructed
+    for (int i = 0; i < mins.size(); i++) {
+      insert(mins.get(i), maxs.get(i), values.get(i));
+    }
+  }
+
+  private void insert(long min, long max, Value value) {
+    root = insert(root, new Interval(min, max), value);
+  }
+
+  private Node insert(Node node, Interval interval, Value value) {
+    if (node == null) {
+      return new Node(interval, value);
+    }
+
+    int cmp = interval.compareTo(node.interval);
+    if (cmp == 0) {
+      node.valueList.add(value);
+      return node;
+    } else if (cmp < 0) {
+      node.left = insert(node.left, interval, value);
+    } else {
+      node.right = insert(node.right, interval, value);
+    }
+
+    fixAuxiliaryInfo(node);
+    int balance = getBalance(node);
+    // Balance current subtree
+    if (balance < -1) {
+      if (getBalance(node.right) > 0) {
+        node.right = rotateRight(node.right); // Right-left case
+      }
+      node = rotateLeft(node);
+    } else if (balance > 1) {
+      if (getBalance(node.left) < 0) {
+        node.left = rotateLeft(node.left); // Left-right case
+      }
+      node = rotateRight(node);
+    }
+    return node;
+  }
+
+  // Get all intervals which intersect with input range [min, max]
+  public List<Value> searchAll(long min, long max) {
+    return searchAll(new Interval(min, max));
+  }
+
+  public List<Value> searchAll(Interval interval) {
```

Comment:
This API is a bit non-intuitive. Essentially what it is doing is given an interval, it looks for all overlapping intervals in the BST and for all those overlapping intervals, returns the corresponding list of segments? At least a javadoc with details on semantics is definitely needed